pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,197 @@
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	kube_handler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	kube_reconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
+	kube_source "sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+)
+
+const nodeReadinessTaintKey = "NodeReadiness"
+const nodeIndexField = "spec.nodeName"
+const parentAppLabel = "parent.app"
+
+type CniNodeTaintReconciler struct {
+	kube_client.Client
+	Log logr.Logger
+}
+
+func (r *CniNodeTaintReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("node", req.NamespacedName)
+	log.V(1).Info("event received")
+
+	// Fetch the Node instance
+	node := &kube_core.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.Error(err, "node not found")
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch node")
+		return kube_ctrl.Result{}, err
+	}
+	log.Info("node successfully fetched")
+
+	kubeSystemPods := &kube_core.PodList{}
+	namespaceOption := kube_client.InNamespace("kube-system")
+	matchingFields := kube_client.MatchingFields{nodeIndexField: node.Name}
+	matchingLabels := kube_client.MatchingLabels{parentAppLabel: metadata.KumaCNI}
+	if err := r.Client.List(ctx, kubeSystemPods, namespaceOption, matchingFields, matchingLabels); err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err := r.updateTaints(ctx, log, node, kubeSystemPods.Items)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update node taints")
+	}
+
+	return kube_ctrl.Result{}, err
+}
+
+func (r *CniNodeTaintReconciler) updateTaints(ctx context.Context, log logr.Logger, node *kube_core.Node, pods []kube_core.Pod) error {
+	if hasTaint(node) {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		node.Spec.Taints = removeTaint(node.Spec.Taints, taintIndex)
+	}
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func removeTaint(s []kube_core.Taint, index int) []kube_core.Taint {
```

Comment:
Nit: This gets down to tastes but this codebase has a tendancy to prefer longer methods over small functions.
For example, it feels like this function could be inlined with a comment.

I think @jakubdyszkiewicz is quite in agreement with this blogpost: https://copyconstruct.medium.com/small-functions-considered-harmful-91035d316c29

Reply:
For me this function is just overcoming a shortcoming of golang of not having a proper stdlib. I can just read the name and know what it does without the need to decipher what `append(s[:index], s[index+1:]...)` means. It's fine when there is 1 thing like that but golang seems to be missing quite a lot.

I can use `slices.delete` which is [just what](https://cs.opensource.google/go/x/exp/+/master:slices/slices.go;l=157?q=slices.delete&ss=go%2Fx%2Fexp) I do below.

Reply:
Big agree on go seemingly being engineered to be as unreadable as possible and small functions improving it. Though personally I think here we could keep the `Update` call in `updateTaints` and create a `removeTaintIfPresent` to keep the separation between pure and effectful code as much as possible.

Reply:
In the risk of bikeshedding I find:

```go
	if taintIndex >= 0 {
                 // remove taint
		node.Spec.Taints = append(node.Spec.Taints[:taintIndex], node.Spec.Taints[taintIndex+1:]...)
	}
```

Easier to read than:


```go
	if taintIndex >= 0 {
		node.Spec.Taints = removeTaint(node.Spec.Taints, taintIndex)
	}
// ...

func removeTaint(s []kube_core.Taint, index int) []kube_core.Taint {
       return append(s[:index], s[index+1:]...)
}
```

I feel it respects the flow of reading more easily. Again no strong feelings here.

Reply:
I refactored it a bit using your's and Mike's suggestions, let me know if it's better now